signer/src/error.rs

```diff
@@ -3,11 +3,31 @@ use std::borrow::Cow;
 
 use blockstack_lib::types::chainstate::StacksBlockId;
 
-use crate::{codec, ecdsa, network};
+use crate::{codec, context::SignerSignal, ecdsa, network};
 
 /// Top-level signer error
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+    /// Error broadcasting a message to the application channel
+    #[error("Error broadcasting message to application channel: {0}")]
+    BroadcastSendError(#[from] tokio::sync::broadcast::error::SendError<SignerSignal>),
+
+    /// Error receiving a message from the application channel
+    #[error("Error receiving message from application channel: {0}")]
+    BroadcastRecvError(#[from] tokio::sync::broadcast::error::RecvError),
```

Comment:
Both of these error types should be handled by the logic and not propagated imo. For example, the send error should only occur if the receiver side is dropped, which should only happen during shutdown and can be used as a signal to terminate the specific sub-channel instead of having a dedicated termination channel.

Reply:
My two cents here is that while I'm not against handling shutdown signals explicitly, I think it can be pretty nice to handle it implicitly as well. If you have a handle to a subtask in the form of a send or receiver channel (or both), it makes sense for the subtask to gracefully shut down once either of those are dropped, similar to how normal structures are deallocated when any references to them are dropped.

I do see that there might be scenarios where some extra sophisticated shutdown management is needed, but in my experience you can get very far by this simple rule.

Reply:
The issue I see is that it explicitly requires a `send` to occur in order to detect that all receivers have been dropped -- otherwise the loop will continue indefinitely waiting for a command to broadcast a p2p message (thus blocking shutdown). The channel lives as long as the `SignerContext` (as it holds a reference to a `Sender` to be able to give out new subscriptions), so the channel itself won't be dropped until the end of the program.

Reply:
Oh right, similar to how cyclic references are never garbage collected. Fair enough, then I see the point in having the explicit termination signal.